File: python-algorithm/ds_8_matrix/L54-m.py
# -*- coding: utf-8 -*-


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res = []
        m, n = len(matrix), len(matrix[0])
        up, down, left, right = 0, m - 1, 0, n - 1
        while True:
            for i in range(left, right + 1):
                res.append(matrix[up][i])
            up += 1
            if up > down: break

            for i in range(up, down + 1):
                res.append(matrix[i][right])
            right -= 1
            if right < left: break

            # 如果range的起始值大于结束值，后面必须指定步长
            for i in range(right, left - 1, -1):
                res.append(matrix[down][i])
            down -= 1
            if down < up: break

            for i in range(down, up - 1, -1):
                res.append(matrix[i][left])
            left += 1
            if left > right: break

        return res

File: python-algorithm/ds_8_matrix/L54-m.cpp


//螺旋矩阵

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        int m = matrix.size();
        if(m == 0)
            return res;
        int n = matrix[0].size();

        // 上下左右
        int u = 0;
        int d = m-1;
        int l = 0;
        int r = n-1;
        int num = 1;

        while(true){
            // 上边界
            for(int i=l; i <= r; i++) res.push_back(matrix[u][i]);
            //if (u++ >= d) break; //转向时候 ↓ 移一位不越界
            if(u >= d) break;
            u = u + 1;
            // 右边界
            for(int i=u; i <= d; i++) res.push_back(matrix[i][r]);
            //if (r-- <= l) break; //转向时候 ← 移一位不越界
            if(r <= l ) break;
            r =  r - 1;
            // 下边界
            for(int i=r; i >= l; i--) res.push_back(matrix[d][i]);
            //if (d-- <= u) break; //转向时候 ↑ 移一位不越界
            if(d <= u) break;
            d = d - 1;
            // 左边界
            for(int i=d; i >= u; i--) res.push_back(matrix[i][l]);
            //if (l++ >= r) break; //转向时候 → 移一位不越界
            if(l >= r) break ;
            l = l + 1 ;
        }
        return res;
    }
};


int main(){
    return 0;
}